File: control-plane/pkg/types/status.go
package types

import "strings"

// ExecutionStatus represents the canonical set of execution statuses supported by the system.
type ExecutionStatus = string

const (
	ExecutionStatusUnknown   ExecutionStatus = "unknown"
	ExecutionStatusPending   ExecutionStatus = "pending"
	ExecutionStatusQueued    ExecutionStatus = "queued"
	ExecutionStatusWaiting   ExecutionStatus = "waiting"
	ExecutionStatusRunning   ExecutionStatus = "running"
	ExecutionStatusSucceeded ExecutionStatus = "succeeded"
	ExecutionStatusFailed    ExecutionStatus = "failed"
	ExecutionStatusCancelled ExecutionStatus = "cancelled"
	ExecutionStatusTimeout   ExecutionStatus = "timeout"
)

var canonicalExecutionStatuses = map[ExecutionStatus]struct{}{
	ExecutionStatusUnknown:   {},
	ExecutionStatusPending:   {},
	ExecutionStatusQueued:    {},
	ExecutionStatusWaiting:   {},
	ExecutionStatusRunning:   {},
	ExecutionStatusSucceeded: {},
	ExecutionStatusFailed:    {},
	ExecutionStatusCancelled: {},
	ExecutionStatusTimeout:   {},
}

var executionStatusAliases = map[string]ExecutionStatus{
	"success":           ExecutionStatusSucceeded,
	"successful":        ExecutionStatusSucceeded,
	"completed":         ExecutionStatusSucceeded,
	"complete":          ExecutionStatusSucceeded,
	"done":              ExecutionStatusSucceeded,
	"ok":                ExecutionStatusSucceeded,
	"error":             ExecutionStatusFailed,
	"failure":           ExecutionStatusFailed,
	"errored":           ExecutionStatusFailed,
	"canceled":          ExecutionStatusCancelled,
	"cancel":            ExecutionStatusCancelled,
	"timed_out":         ExecutionStatusTimeout,
	"wait":              ExecutionStatusQueued,
	"awaiting_approval": ExecutionStatusWaiting,
	"awaiting_human":    ExecutionStatusWaiting,
	"approval_pending":  ExecutionStatusWaiting,
	"in_progress":       ExecutionStatusRunning,
	"processing":        ExecutionStatusRunning,
}

// NormalizeExecutionStatus maps arbitrary status strings onto the canonical execution statuses used by the AgentField platform.
// Unknown or unsupported statuses resolve to ExecutionStatusUnknown.
func NormalizeExecutionStatus(status string) string {
	normalized := ExecutionStatus(strings.ToLower(strings.TrimSpace(status)))
	if normalized == "" {
		return string(ExecutionStatusUnknown)
	}
	if _, ok := canonicalExecutionStatuses[normalized]; ok {
		return string(normalized)
	}
	if mapped, ok := executionStatusAliases[string(normalized)]; ok {
		return string(mapped)
	}
	return string(ExecutionStatusUnknown)
}

// IsTerminalExecutionStatus reports whether the provided status string represents a terminal execution state.
func IsTerminalExecutionStatus(status string) bool {
	switch NormalizeExecutionStatus(status) {
	case string(ExecutionStatusSucceeded), string(ExecutionStatusFailed), string(ExecutionStatusCancelled), string(ExecutionStatusTimeout):
		return true
	default:
		return false
	}
}

File: sdk/python/agentfield/status.py
"""Canonical execution status utilities for the AgentField SDK."""

from __future__ import annotations

from typing import Optional, Set, Tuple

CANONICAL_STATUSES: Tuple[str, ...] = (
    "pending",
    "queued",
    "waiting",
    "running",
    "succeeded",
    "failed",
    "cancelled",
    "timeout",
    "unknown",
)

CANONICAL_STATUS_SET: Set[str] = set(CANONICAL_STATUSES)

_STATUS_ALIASES = {
    "success": "succeeded",
    "successful": "succeeded",
    "completed": "succeeded",
    "complete": "succeeded",
    "done": "succeeded",
    "ok": "succeeded",
    "error": "failed",
    "failure": "failed",
    "errored": "failed",
    "canceled": "cancelled",
    "cancel": "cancelled",
    "timed_out": "timeout",
    "wait": "queued",
    "awaiting_approval": "waiting",
    "awaiting_human": "waiting",
    "approval_pending": "waiting",
    "in_progress": "running",
    "processing": "running",
}

TERMINAL_STATUSES: Set[str] = {"succeeded", "failed", "cancelled", "timeout"}


def normalize_status(status: Optional[str]) -> str:
    """Return the canonical representation of a status string."""

    if status is None:
        return "unknown"

    normalized = status.strip().lower()
    if not normalized:
        return "unknown"

    if normalized in CANONICAL_STATUS_SET:
        return normalized

    return _STATUS_ALIASES.get(normalized, "unknown")


def is_terminal(status: Optional[str]) -> bool:
    """Return True if the provided status represents a terminal state."""

    return normalize_status(status) in TERMINAL_STATUSES


__all__ = [
    "CANONICAL_STATUSES",
    "CANONICAL_STATUS_SET",
    "TERMINAL_STATUSES",
    "normalize_status",
    "is_terminal",
]
